server/lib/InstanceManager.php

```diff
@@ -0,0 +1,238 @@
+<?php
+
+
+namespace LookupServer;
+
+use PDO;
+use Psr\Http\Message\ResponseInterface as Response;
+use Psr\Http\Message\ServerRequestInterface as Request;
+
+
+class InstanceManager {
+
+
+	/** @var PDO */
+	private $db;
+
+	/** @var SignatureHandler */
+	private $signatureHandler;
+
+
+	/**
+	 * InstanceManager constructor.
+	 *
+	 * @param PDO $db
+	 * @param SignatureHandler $signatureHandler
+	 */
+	public function __construct(PDO $db, SignatureHandler $signatureHandler) {
+		$this->db = $db;
+		$this->signatureHandler = $signatureHandler;
+	}
+
+
+	public function insert(string $instance) {
+		$stmt = $this->db->prepare('SELECT id, instance, timestamp FROM instances WHERE instance=:instance');
+		$stmt->bindParam(':instance', $instance, PDO::PARAM_STR);
+		$stmt->execute();
+
+		$data = $stmt->fetch();
+		if ($data === false) {
+			$time = time();
+			$insert = $this->db->prepare(
+				'INSERT INTO instances (instance, timestamp) VALUES (:instance, FROM_UNIXTIME(:timestamp))'
+			);
+			$insert->bindParam(':instance', $instance, PDO::PARAM_STR);
+			$insert->bindParam(':timestamp', $time, PDO::PARAM_INT);
+
+			$insert->execute();
+		}
+	}
+
+
+	/**
+	 * let Nextcloud servers obtains the full list of registered instances in the global scale scenario
+	 * If result is empty, sync from the users list
+	 *
+	 * @param Request $request
+	 * @param Response $response
+	 *
+	 * @return Response
+	 */
+	public function getInstances(Request $request, Response $response): Response {
+		try {
+			$source = $this->signatureHandler->verifyRequest($request);
```

Comment:
You should probably have some shared secret here between the nodes and the lookupserver. So that only if that is set you can actually get this.

Reply:
And it should only work in GS mode anyways I think

Reply:
@rullzer ok, I removed the signature check and replaced it with the authKey and limited the request when GS mode enabled.